scylla/tests/integration/tablets.rs

```diff
@@ -0,0 +1,628 @@
+use std::sync::Arc;
+
+use crate::utils::setup_tracing;
+use crate::utils::test_with_3_node_cluster;
+
+use futures::future::try_join_all;
+use itertools::Itertools;
+use scylla::load_balancing::FallbackPlan;
+use scylla::load_balancing::LoadBalancingPolicy;
+use scylla::load_balancing::RoutingInfo;
+use scylla::prepared_statement::PreparedStatement;
+use scylla::query::Query;
+use scylla::serialize::row::SerializeRow;
+use scylla::test_utils::unique_keyspace_name;
+use scylla::transport::ClusterData;
+use scylla::transport::Node;
+use scylla::transport::NodeRef;
+use scylla::ExecutionProfile;
+use scylla::QueryResult;
+use scylla::{IntoTypedRows, Session};
+
+use scylla_cql::errors::QueryError;
+use scylla_proxy::{
+    Condition, ProxyError, Reaction, ResponseFrame, ResponseOpcode, ResponseReaction, ResponseRule,
+    ShardAwareness, TargetShard, WorkerError,
+};
+
+use tokio::sync::mpsc;
+use tracing::info;
+use uuid::Uuid;
+
+#[derive(scylla::FromRow)]
+struct SelectedTablet {
+    last_token: i64,
+    replicas: Vec<(Uuid, i32)>,
+}
+
+struct Tablet {
+    first_token: i64,
+    last_token: i64,
+    replicas: Vec<(Arc<Node>, i32)>,
+}
+
+async fn get_tablets(session: &Session, ks: String, table: String) -> Vec<Tablet> {
+    let cluster_data = session.get_cluster_data();
+    let endpoints = cluster_data.get_nodes_info();
+    for endpoint in endpoints.iter() {
+        info!(
+            "Endpoint id: {}, address: {}",
+            endpoint.host_id,
+            endpoint.address.ip()
+        );
+    }
+
+    let selected_tablets_rows = session.query(
+        "select last_token, replicas from system.tablets WHERE keyspace_name = ? and table_name = ? ALLOW FILTERING",
+        &(ks.as_str(), table.as_str())).await.unwrap().rows.unwrap();
+
+    let mut selected_tablets = selected_tablets_rows
+        .into_typed::<SelectedTablet>()
+        .map(|x| x.unwrap())
+        .collect::<Vec<_>>();
+    selected_tablets.sort_unstable_by(|a, b| a.last_token.cmp(&b.last_token));
+
+    let mut tablets = Vec::new();
+    let mut first_token = i64::MIN;
+    for tablet in selected_tablets {
```

Comment:
There will still be `mut first_token` - it will be more confusing if we are using functional APIs that mutate some outside variable.
But it actually can be done with `fold` - I think it is one of the rare cases where it is a proper tool.
There is still `tablets.push`, but there are not `mut` vars outside fold call.